In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
import csv

In [3]:
path = "C:/Users/apujari/Desktop/sales.csv"
sales_original = pd.read_csv(path, ",")
sales_original

Unnamed: 0     ID   Prod  Quant     Val  Insp
0                1     v1     p1    182    1665  unkn
1                2     v2     p1   3072    8780  unkn
2                3     v3     p1  20393   76990  unkn
3                4     v4     p1    112    1100  unkn
4                5     v3     p1   6164   20260  unkn
5                6     v5     p2    104    1155  unkn
6                7     v6     p2    350    5680  unkn
7                8     v7     p2    200    4010  unkn
8                9     v8     p2    233    2855  unkn
9               10     v9     p2    118    1175  unkn
10              11    v10     p2    233    1845  unkn
11              13    v12     p3    108    1055  unkn
12              14    v13     p3   1775    9285  unkn
13              15    v14     p3    304   28300  unkn
14              16    v13     p3   5211   27900  unkn
15              17    v13     p4   9452   35540  unkn
16              18    v15     p4    147    1555  unkn
17              19     v6     p4    366    6300  unkn
18              20    v13     p4   1836   12205  unkn
19              21    v16     p4   1233    6735  unkn
20              22    v17     p4   2469   17855  unkn
21              23    v18     p4   5332   53695  unkn
22              25    v20     p5   6170  153730  unkn
23              26    v20     p5    180    3175  unkn
24              27    v21     p5    985    6595  unkn
25              28    v22     p5    120    1060  unkn
26              29    v20     p5    361    7950  unkn
27              30    v23     p5    667   11355  unkn
28              31    v24     p5    933    1060  unkn
29              32    v25     p5  11203   54465  unkn
...            ...    ...    ...    ...     ...   ...
401116      411789  v4217  p4125    113    2455  unkn
401117      411790  v5636  p4125    106    1030  unkn
401118      411791  v1134  p4125    101    1025  unkn
401119      411792  v2645  p4125    294    4350  unkn
401120      411793  v5400  p4125    175    3795  unkn
401121      411794  v1705  p4125    216    8200  unkn
401122      411795  v4834  p4125    102    1065  unkn
401123      411796  v1677  p4125    122    1710  unkn
401124      411797  v2670  p4125    116    2070  unkn
401125      411798  v1720  p4125    263   12165  unkn
401126      411799  v2659  p4125    129    2245  unkn
401127      411800  v2236  p4125    283   12195  unkn
401128      411801  v1684  p4125    113    9230  unkn
401129      411802  v5918  p4126    102    1055  unkn
401130      411803  v5918  p4126    133    1520  unkn
401131      411804  v5918  p4126   2261   52855  unkn
401132      411805  v5918  p4126   2994   27675  unkn
401133      411806  v2036  p4126    391    8900  unkn
401134      411807  v2682  p4126    100    1015  unkn
401135      411808  v1783  p4126    172    4170  unkn
401136      411809  v4117  p4126    104    1385  unkn
401137      411810  v1784  p4126    105    1465  unkn
401138      411811  v2248  p4126    144    5545  unkn
401139      411812  v1783  p4126    438    7940  unkn
401140      411813  v4117  p4126    287    8365  unkn
401141      411814  v5918  p4126    113    9330  unkn
401142      411815  v2540  p4126    118    3355  unkn
401143      411816  v5896  p4127    143    1835  unkn
401144      411817  v1796  p4127    102    1445  unkn
401145      411818  v2540  p4127    106    1275  unkn

[401146 rows x 6 columns]

In [4]:
## Val and Quant are the only fields with missing data. 
## After examining both ok and froaud datasets it is observed that both 
## have similar amounts of missing data.
## A valid reason may be that the sale was not complete. 
## So dropping the records from analysis would be better.

sales = sales_original.dropna()
sales.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 387010 entries, 0 to 401145
Data columns (total 6 columns):
Unnamed: 0    387010 non-null int64
ID            387010 non-null object
Prod          387010 non-null object
Quant         387010 non-null float64
Val           387010 non-null float64
Insp          387010 non-null object
dtypes: float64(2), int64(1), object(3)
memory usage: 16.2+ MB


# Sanity check/ Explore Data

In [5]:
## Create a plot of OK and Fraud points excluding extreme volues
## to see how the OK and Fraud points relate to each other.
## Note that analysis showed that both ok and fraud datasets have extreme values.

salesOK = sales[sales.Insp == 'ok']
testOK = salesOK[(salesOK.Quant < 5000) & (salesOK.Val < 60000)]
salesF = sales[sales.Insp == 'fraud']
testF = salesF[(salesF.Quant < 5000) & (salesF.Val < 60000)]

plt.scatter(testOK.Quant, testOK.Val, color = 'blue')

plt.scatter(testF.Quant, testF.Val, color = 'red')


In [6]:
plt.show()

# Please check the attached file for the scatterplot of ok and fraud points.

### The plot shows no obvious relationship between Val and Quant. Clustering and Linear models won't be useful. This plot shows that the ratiobetween Quant and Val for OK sales can take almost any value but for most fraud sales, the ratio is extreme, either very high or very low. If the Price of a product is Val / Quant. In many draud cases, the price is very high (Val for Quant was too high) or price is very low (Quant for Val to high). This gave me the idea to add a Price column to the sales dataset and my analysis concentrated on finding extreme values of Price.

In [8]:
sales["Price"] = sales.Val / sales.Quant
sales["Price"]

C:\Anaconda\lib\site-packages\IPython\kernel\__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


0          9.148352
1          2.858073
2          3.775315
3          9.821429
4          3.286827
5         11.105769
6         16.228571
7         20.050000
8         12.253219
9          9.957627
10         7.918455
11         9.768519
12         5.230986
13        93.092105
14         5.354059
15         3.760051
16        10.578231
17        17.213115
18         6.647603
19         5.462287
20         7.231673
21        10.070330
22        24.915721
23        17.638889
24         6.695431
25         8.833333
26        22.022161
27        17.023988
28         1.136120
29         4.861644
            ...    
401116    21.725664
401117     9.716981
401118    10.148515
401119    14.795918
401120    21.685714
401121    37.962963
401122    10.441176
401123    14.016393
401124    17.844828
401125    46.254753
401126    17.403101
401127    43.091873
401128    81.681416
401129    10.343137
401130    11.428571
401131    23.376824
401132     9.243487
401133    22.762148
401134    10.150000


### Summary

#### The main focus of the anomaly detection methods that I try, will be based on the idea that Price (Val/Quant) will by consistent for each product. Fraud will be detected when the Price for a sale is higher or lower than expected based on the Price of other sales of the same product.

## Unsupervised Approaches

### (1) Student's t using Prices from sales of same Prod

In [9]:
## Create a random sample of 500 records from the fraud label and 500 from the ok label.
test = pd.DataFrame(sales[sales.Insp == 'fraud'].sample(500))
test2 = pd.DataFrame(sales[sales.Insp == 'ok'].sample(500))

test.append(test2)
test = test.append(test2)
test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 225106 to 9076
Data columns (total 7 columns):
Unnamed: 0    1000 non-null int64
ID            1000 non-null object
Prod          1000 non-null object
Quant         1000 non-null float64
Val           1000 non-null float64
Insp          1000 non-null object
Price         1000 non-null float64
dtypes: float64(3), int64(1), object(3)
memory usage: 50.8+ KB


In [10]:
## Create a subset of the sales data using only sales for unkn products
unSupervised = sales[sales.Insp == 'unkn']
unSupervised.describe()

Unnamed: 0            Quant             Val          Price
count  371464.00000    371464.000000   371464.000000  371464.000000
mean   206827.86915      4263.115303    13018.143508      19.116699
std    118411.54454     63102.643921    61784.191164      87.361592
min         1.00000       100.000000     1005.000000       0.006147
25%    104431.75000       107.000000     1400.000000       8.512593
50%    207075.50000       166.000000     2770.000000      11.929825
75%    309163.25000       683.000000     8490.000000      18.952381
max    411818.00000  11011711.000000  4308620.000000   21167.902098

In [11]:
# Add two columns to the test dataframe = TureInsp and PredInsp for
# later analysis of the method
InspCat = {'ok' : 1, 'fraud' : 2, 'unkn' : 0}
test["TrueInsp"] = test["Insp"]
test.TrueInsp = test["TrueInsp"].apply(lambda x: InspCat[x])
test["PredInsp"] = 0
test.head()

Unnamed: 0     ID   Prod  Quant      Val   Insp       Price  TrueInsp  \
225106      231415  v5772  p1367    205    47745  fraud  232.902439         2   
38035        39177   v588   p955  52877  2201595  fraud   41.636156         2   
31330        32111   v460  p3997  31333     1290  fraud    0.041171         2   
149810      153569  v5077  p2447    144    84960  fraud  590.000000         2   
252258      258971  v2991  p2962  33760     1460  fraud    0.043246         2   

        PredInsp  
225106         0  
38035          0  
31330          0  
149810         0  
252258         0

In [12]:
# For every sale in the test dataset, compare the price of the sale 
# to the price of the sales with the same product id, Prod
# use a student's t value and an interval of 80%
from scipy.stats import t

for index in test.index:
    prod = test.Prod.loc[index]
    price = test.Price.loc[index]
    prices = unSupervised[unSupervised.Prod == prod]["Price"]
    tvalue = (price - prices.values.mean()) / prices.values.std()
    df = len(prices) - 1
    interval = t.interval(.80, df)
    if price < interval[0] or price >interval[1]:
        test["PredInsp"].loc[index] = 2
    else:
        test["PredInsp"].loc[index] = 1
test.head()

C:\Anaconda\lib\site-packages\pandas\core\indexing.py:115: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


Unnamed: 0     ID   Prod  Quant      Val   Insp       Price  TrueInsp  \
225106      231415  v5772  p1367    205    47745  fraud  232.902439         2   
38035        39177   v588   p955  52877  2201595  fraud   41.636156         2   
31330        32111   v460  p3997  31333     1290  fraud    0.041171         2   
149810      153569  v5077  p2447    144    84960  fraud  590.000000         2   
252258      258971  v2991  p2962  33760     1460  fraud    0.043246         2   

        PredInsp  
225106         2  
38035          2  
31330          1  
149810         2  
252258         1

In [13]:
## Evaluate the method by determining the precision and recall
from sklearn.metrics import precision_recall_fscore_support

results = precision_recall_fscore_support(test.TrueInsp, test.PredInsp)
results

(array([ 0.12796209,  0.40050697]),
 array([ 0.054,  0.632]),
 array([ 0.07594937,  0.49030256]),
 array([500, 500]))

### The above results show that precision of identifying the Fraud sales was 39.5% and the recall was 61.4%. Overall, the results were mislabelled. The majoruty the true fraud sales were found but it wasn't a high majority.

### (2) Student's t using Prices from same Prod and ID

#### The documentation states that the proce of the product can vary depending on the sales person and the quantity of the sale. Repeating the above method by restricting the orher prices considered by salesperson as well as product.

In [14]:
for index in test.index:
    prod = test.Prod.loc[index]
    seller = test.ID.loc[index]
    price = test.Price.loc[index]
    prices = unSupervised[(unSupervised.Prod == prod) & (unSupervised.ID == seller)]["Price"]
    df = len(prices) - 1
    interval = t.interval(.80, df)
    if price < interval[0] or price > interval[1]:
        test["PredInsp"].loc[index] = 2
    else:
        test["PredInsp"].loc[index] = 1
test.head()

Unnamed: 0     ID   Prod  Quant      Val   Insp       Price  TrueInsp  \
225106      231415  v5772  p1367    205    47745  fraud  232.902439         2   
38035        39177   v588   p955  52877  2201595  fraud   41.636156         2   
31330        32111   v460  p3997  31333     1290  fraud    0.041171         2   
149810      153569  v5077  p2447    144    84960  fraud  590.000000         2   
252258      258971  v2991  p2962  33760     1460  fraud    0.043246         2   

        PredInsp  
225106         2  
38035          2  
31330          1  
149810         2  
252258         1

In [15]:
results = precision_recall_fscore_support(test.TrueInsp, test.PredInsp)
results

(array([ 0.42627346,  0.45614035]),
 array([ 0.318,  0.572]),
 array([ 0.36426117,  0.50754215]),
 array([500, 500]))

### The above results show that precision of identifying the Fraud sales was 43% and the recall was 54.2%. The precision and recall for 'ok' sales were much improved however the interpretation can go like this time as poorer than the precious because the recall of the fraud sales were lower.

## Summary

### The recall of the fraud cases was actually a bit better when we used the Price values from sales records without taking the ID of the record into account.

# Semisupervised Techniques

## (1) Use sales records labeled 'ok'

In [16]:
### There are sales records that hace been labelled 'ok' or 'fraud'/
### The 'ok' records could be used ub seni-supervised methods as a basis 
### for establishing 'normal' data.

allProducts = set(sales.Prod.unique())
okProducts = set(sales[sales.Insp == 'ok']["Prod"].unique())
len(allProducts & okProducts), len(okProducts), len(allProducts)

(797, 797, 4546)

### The above results show that there are significantly more products in general
### that have been labelled 'ok'. Perhaps some products are no longer sold.


In [17]:
semiSuper = sales[sales.Insp == 'ok']
semiSuper.describe()

Unnamed: 0            Quant             Val         Price
count   14347.000000     14347.000000    14347.000000  14347.000000
mean   175715.334286     35796.608699    61238.326828     33.797279
std    127397.502763    340108.861246   148928.636851    144.748862
min        53.000000       100.000000     1005.000000      0.005561
25%     52608.000000       109.000000     1255.000000      7.522293
50%    164248.000000       433.000000    13915.000000     10.850000
75%    288946.500000      5848.000000    57565.000000     26.170583
max    411478.000000  14645544.000000  2978705.000000   5255.342466

In [18]:
## Create test data from the sales records labelled fraud

test = pd.DataFrame(sales[sales.Insp == 'fraud'])

## Add two columns to the test daraframe: TrueInsp and PredInsp
## for later analusis of the method.
InspCat = {'ok' : 1, 'fraud' : 2, 'unkn' : 0}
test["TrueInsp"] = test["Insp"]
test.TrueInsp = test["TrueInsp"].apply(lambda x: InspCat[x])
test["PredInsp"] = 0
test.head()

Unnamed: 0    ID  Prod  Quant     Val   Insp       Price  TrueInsp  \
379          391   v68   p59    111   23000  fraud  207.207207         2   
531          564  v108  p109  34611   12485  fraud    0.360723         2   
959         1019  v169  p216   1986  132800  fraud   66.868077         2   
1053        1113  v194  p226    114    4515  fraud   39.605263         2   
1054        1114  v194  p226    213   11745  fraud   55.140845         2   

      PredInsp  
379          0  
531          0  
959          0  
1053         0  
1054         0

In [19]:
for index in test.index:
    prod = test.Prod.loc[index]
    price = test.Price.loc[index]
    prices = semiSuper[semiSuper.Prod == prod]["Price"]
    tvalue = (prices - prices.values.mean()) / prices.values.std()
    df = len(prices) - 1
    interval = t.interval(0.80, df)
    if price < interval[0] or price > interval[1]:
        test["PredInsp"].loc[index] = 2
    else:
        test["PredInsp"].loc[index] = 1
test.head()        

Unnamed: 0    ID  Prod  Quant     Val   Insp       Price  TrueInsp  \
379          391   v68   p59    111   23000  fraud  207.207207         2   
531          564  v108  p109  34611   12485  fraud    0.360723         2   
959         1019  v169  p216   1986  132800  fraud   66.868077         2   
1053        1113  v194  p226    114    4515  fraud   39.605263         2   
1054        1114  v194  p226    213   11745  fraud   55.140845         2   

      PredInsp  
379          2  
531          1  
959          2  
1053         2  
1054         2

In [20]:
results = precision_recall_fscore_support(test.TrueInsp, test.PredInsp)
results

C:\Anaconda\lib\site-packages\sklearn\metrics\classification.py:960: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples.
  'recall', 'true', average, warn_for)


(array([ 0.,  1.]),
 array([ 0.        ,  0.58548791]),
 array([ 0.        ,  0.73855865]),
 array([   0, 1199]))

### Here the results show that the recall was 58.5%.
### Next we will try combining the unlabelled data with the 'ok' data.

### (2) Combine data labeled ok and unlabelled data

In [22]:
# Combine the ok and unkn sales records
semiSuper = sales[(sales.Insp != 'fraud')]
for index in test.index:
    prod = test.Prod.loc[index]
    price = test.Price.loc[index]
    prices = semiSuper[semiSuper.Prod == prod]["Price"]
    tvalue = (price - prices.values.mean())/prices.values.std()
    df = len(prices) - 1
    interval = t.interval(0.80, df)
    if price < interval[0] or price > interval[1]:
        test["PredInsp"].loc[index] = 2
    else:
        test["PredInsp"].loc[index] = 1
test.head()

Unnamed: 0    ID  Prod  Quant     Val   Insp       Price  TrueInsp  \
379          391   v68   p59    111   23000  fraud  207.207207         2   
531          564  v108  p109  34611   12485  fraud    0.360723         2   
959         1019  v169  p216   1986  132800  fraud   66.868077         2   
1053        1113  v194  p226    114    4515  fraud   39.605263         2   
1054        1114  v194  p226    213   11745  fraud   55.140845         2   

      PredInsp  
379          2  
531          1  
959          2  
1053         2  
1054         2

In [23]:
results = precision_recall_fscore_support(test.TrueInsp, test.PredInsp)
results

(array([ 0.,  1.]),
 array([ 0.        ,  0.63469558]),
 array([ 0.        ,  0.77653061]),
 array([   0, 1199]))

### The results show that the recall is almost 63.5%

# Summary

### As with the unsupervised approach, we got better results when a larger set of data is used.


## Supervised Techniques
### Because of the existence of labelled Fraud and OK sales records it is tempting to use a supervised technique to serach for outliers. The problem is that there are over 400 unique products and the labelled fraud data accounts for less than 800 of them. There isn't enough coverage of possivle sales fraud to use one of these techniques. It might be possible that certain products hace a high risk of fraudulent sales. Perhaps for those products a proximity technique that uses the Price to classify the sale as Fraud or OK.



## Other Creative Strategies

### Throughout the unsupervised and semi-supervised techniques that we used here, the recall was around 60%. We would like to sample some of the fraud sales that were not recalled to see what is unique about them and think about a method that would detect them. One idea is to do with extreme values of both Val and Quant. When we first examined the data, we found that there were extreme values in both the fraud and ok sales so extreme values weren't alarming but the Fraud data had more extreme values. This analysis concentrated on the ratio of Val to Quant. That ratio could be valid even though the values of Quant or Val are extreme. It might be worth exploring approaches that look for extreme values for Quant or Val among the same Prod. 
